src/Data/Singletons/Util.hs

```diff
@@ -29,6 +30,13 @@ import Data.Foldable
 import Data.Traversable
 import Data.Generics
 import Control.Monad.Fail ( MonadFail )
+import GHC.Exts (Constraint)
+
+-- | Force GHC to unify the kinds of @a@ and @b@. Note that @SameKind a b@ is
+-- different from @KindOf a ~ KindOf b@ in that the former makes the kinds
+-- unify immediately, whereas the latter is a proposition that GHC considers
+-- as possibly false.
+type SameKind (a :: k) (b :: k) = (() :: Constraint)
```

Comment:
I think this belongs better in `Data.Singletons`, as it will appear in the generated code.